processing/src/main/java/org/apache/druid/segment/ColumnarValueUnnestCursor.java

```diff
@@ -0,0 +1,270 @@
+          {
+            baseColumSelectorFactory.makeColumnValueSelector(columnName).inspectRuntimeShape(inspector);
+          }
+
+          @Override
+          public boolean isNull()
+          {
+            return getObject() == null;
+          }
+
+          @Nullable
+          @Override
+          public Object getObject()
+          {
+            if (!unnestListForCurrentRow.isEmpty()) {
+              if (allowSet == null || allowSet.isEmpty()) {
+                return unnestListForCurrentRow.get(index);
+              } else if (allowSet.contains((String) unnestListForCurrentRow.get(index))) {
+                return unnestListForCurrentRow.get(index);
+              }
+            }
+            return null;
+          }
+
+          @Override
+          public Class classOfObject()
+          {
+            return Object.class;
+          }
+        };
+      }
+
+      @Nullable
+      @Override
+      public ColumnCapabilities getColumnCapabilities(String column)
+      {
+        if (!outputName.equals(columnName)) {
+          baseColumSelectorFactory.getColumnCapabilities(column);
+        }
+        return baseColumSelectorFactory.getColumnCapabilities(columnName);
+      }
+    };
+  }
+
+  @Override
+  public DateTime getTime()
+  {
+    return baseCursor.getTime();
+  }
+
+  @Override
+  public void advance()
+  {
+    advanceUninterruptibly();
+    BaseQuery.checkInterrupted();
+  }
+
+  @Override
+  public void advanceUninterruptibly()
+  {
+    do {
+      advanceAndUpdate();
+    } while (matchAndProceed());
+  }
+
+  @Override
+  public boolean isDone()
+  {
+    if (needInitialization && baseCursor.isDone() == false) {
+      initialize();
+    }
+    return baseCursor.isDone();
+  }
+
+  @Override
+  public boolean isDoneOrInterrupted()
+  {
+    if (needInitialization && baseCursor.isDoneOrInterrupted() == false) {
+      initialize();
+    }
+    return baseCursor.isDoneOrInterrupted();
+  }
+
+  @Override
+  public void reset()
+  {
+    index = 0;
+    needInitialization = true;
+    baseCursor.reset();
+  }
+
+  @Override
+  public void initialize()
+  {
+    if (columnValueSelector != null) {
+      this.currentVal = this.columnValueSelector.getObject();
+      this.unnestListForCurrentRow = new ArrayList<>();
+      if (currentVal == null) {
+        unnestListForCurrentRow = new ArrayList<>();
+        unnestListForCurrentRow.add(null);
+      } else {
+        if (currentVal instanceof List) {
+          unnestListForCurrentRow = (List<Object>) currentVal;
+        } else if (currentVal.getClass().equals(String.class)) {
+          unnestListForCurrentRow.add(currentVal);
+        }
+      }
+      if (allowSet != null) {
+        if (!allowSet.isEmpty()) {
+          if (!allowSet.contains((String) unnestListForCurrentRow.get(index))) {
+            advance();
+          }
+        }
+      }
+    }
+    needInitialization = false;
+  }
+
+  @Override
+  public void advanceAndUpdate()
```

Comment:
The logic of `initialize` and the logic of `advanceAndUpdate` seem very, very similar to me.  Is there a reason that the initialization cannot just be a pre-emptive `advanceAndUpdate`?